src/GdbConnection.cc

```diff
@@ -1712,6 +1728,172 @@ void GdbConnection::reply_get_thread_list(const vector<GdbThreadId>& threads) {
+    };
+    map<string, library> libraries;
+
+    // iterate through the mappings and search for PE libraries
+    // see wine, packet_query_libraries_cb
+    char buf[0x400];
+    for (auto it = mapping_list.begin(); it != mapping_list.end(); it++) {
+      auto km = *it;
+      bool ok = true;
+      t->read_bytes_helper(km.start(), sizeof(buf), buf, &ok);
+      if (ok) {
+        struct IMAGE_DOS_HEADER* dosheader = (struct IMAGE_DOS_HEADER*)buf;
+        struct IMAGE_NT_HEADERS* ntheader = (struct IMAGE_NT_HEADERS*)(buf + dosheader->e_lfanew);
+        if (dosheader->e_magic == IMAGE_DOS_SIGNATURE &&
+            dosheader->e_lfanew < sizeof(buf)-4 &&
+            ntheader->Signature == IMAGE_NT_SIGNATURE)
+        {
+          if (ntheader->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC ||
+              ntheader->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC)
+          {
+            vector<remote_ptr<void>> segments;
+            struct IMAGE_SECTION_HEADER* sec = IMAGE_FIRST_SECTION(ntheader);
+            for (int i = 0; sec && i < std::max(ntheader->FileHeader.NumberOfSections, (uint16_t)1); ++i) {
+              struct IMAGE_SECTION_HEADER* seci = sec + i;
+              if ((char*)seci > buf + sizeof(buf)) break;
+              segments.push_back(km.start() + seci->VirtualAddress);
+            }
+
+            string name = km.fsname();
+            for (auto it2 = mapping_list.begin(); it2 != mapping_list.end(); it2++) {
+              if (find(segments.begin(), segments.end(), it2->start()) != segments.end()) {
+                if (name.empty() && !it2->fsname().empty()) {
+                  name = it2->fsname();
+                  break;
+                }
+              }
+            }
+
+            // e.g. gdi32.so is prepared by wine to be found as PE image, but gdb does not like it.
+            if (name.substr(name.length() - 3, 3) != ".so") {
+              libraries[name].segments = segments;
+              libraries[name].has_exec = true;
+            }
+          }
+        }
+      }
+    }
+
+    // remove mappings related to PE libraries
+    for (auto m = mapping_list.begin(); m != mapping_list.end(); ) {
+      bool del = false;
+      for (auto lib : libraries) {
+        if (m->fsname() == lib.first) del = true;
+        if (!del) {
+          for (auto seg : lib.second.segments) {
+            if (seg == m->start()) del = true;
+          }
+        }
+      }
+      if (del)
+        m = mapping_list.erase(m);
+      else
+        m++;
+    }
+
+    // search remaining for regular shared objects (look here also into interpreter tables or search ELF signatures?)
+    for (auto it = mapping_list.begin(); it != mapping_list.end(); it++) {
+      auto km = *it;
+      string name = km.fsname();
+      if (!name.empty()) {
+        libraries[name].segments.push_back(km.start());
+        if (km.prot() & PROT_EXEC) {
+          libraries[name].has_exec = true;
+        }
+      }
+    }
```

Comment:
I think we really need to search ld.so's library list if we're going to support this protocol feature. Using the mapping list only is probably not good enough.

Reply:
For now it could be just announced if target_wine is set?

Reply:
Yes, I think that's a good idea.

Reply:
Just for the record, I did not see gdb reqeuest "qXfer:libraries" even when announced when in svr4 mode, just a "qXfer:libraries-svr4".